src/vm/virtualcallstub.cpp

```diff
@@ -3687,11 +3683,16 @@ void DispatchCache::PromoteChainEntry(ResolveCacheElem* elem)
 
     // Remove the element from the chain
     CONSISTENCY_CHECK(curElem->pNext == elem);
-    curElem->pNext = elem->pNext;
+    if(FastInterlockCompareExchangePointer(&curElem->pNext, elem->pNext, elem) != elem)
```

Comment:
There is a race condition here if two threads attempt to promote different elements in the same linked list. For both elements, the FastInterlockCompareExchangePointer succeeds. Then the SetCacheEntry at the end will succeed for just one of them. The one for which it has failed is then inaccessible by walking the list from its head.

Example:
initial structure:
cache[idx]->e0->e1->e2->e3->e4->e5
now thread T1 attempts to promote e2 and T2 attempts to promote e4
so:
* T1 sets e1->next to e3, e2->next to e0
* T2 sets e3->next to e5, e4->next to e0
At this moment, the list looks like
cache[idx]->e0->e1->e3->e5
Now T1 and T2 race for the SetCacheEntry:
* T1 tries to set cache[idx] to e2
* T2 tries to set cache[idx] to e4
Only one of them succeeds. Say it was T1. So the list looks like this:
cache[idx]->e2->e0->e1->e3->e5
Now T2 retries. It goes to retryRemove and then attempts to find e4 in the linked list. But it is not there anymore.

Reply:
@janvorli  I do not think your description matches my intention.

There are two separate retry loops.  One to remove the items from the chain `retryRemove`, and one to insert at head `retryInsert`.  

So flow is intended
* T1 (remove) sets e1->next to e3 (retry to `retryRemove` if fails)
* T1 (insert) set e2->next to e0 and  cache[idx] to e2 (retry to `retryInsert` if fails)

So for your example
* T1 sets e1->next to e3
* T2 sets e3->next to e5
* At this moment, the list looks like
cache[idx]->e0->e1->e3->e5
Now T1 and T2 race for the SetCacheEntry:
* T1 sets e2->next to e0 and tries to set cache[idx] to e2
* T2 sets e4->next to e0 and tries to set cache[idx] to e4
  Only one of them succeeds. Say it was T1. So the list looks like this:
cache[idx]->e2->e0->e1->e3->e5
* Now T2 retries. It goes to retryInsert
 T2 sets e4->next to e2 and tries to set cache[idx] to e4
cache[idx]->e4->e2->e0->e1->e3->e5

I think there may be a race for the parent-child remove race.   
T1 attempts to promote e3 and T2 attempts to promote e2
* T1 sets e2->next to e3->next (e4)
* T2 sets e1->next to e2->next (races e3 or e4)

Proposed solutions
* Use lock only for promote.  Since the race is on remove this should allow lockless insert
* Use temporary chain loop on promoted item
```C++
    prev = Exchange(elem->next, elem);  // Set elem to loop to self
    if (prev == elem)
        return;   // promotion of elem in progress on another thread
retryRemove: 
    parent = find(cache[idx], elem); // May spin if it hits a earlier chain loop
    if (CompareExchange(parent->next, prev) != elem)
        goto retryRemove;
    lastHead = Exchange(cache[idx], elem);  // Set tip to elem
    Exchange(elem->next, lastHead);  // Remove loop
```